North Macedonia municipalities

jsanz opened the pull request:
North Macedonia municipalities.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+ems-file-service+pull-request/6/)

jsanz commented:
@nickpeihl most North Macedonia municipalities OSM relations do not have a wikidata property. Neither wikidata entries have the OSM relation identifier but as per discussion with @nyurik we should add the stable wikidata identifiers to the OSM relationships.

With this Overpass Turbo [query](https://overpass-turbo.eu/s/MqO), I think I have the proper `osm` file to edit and add the identifiers 

```
/* Get Macedonia municipalities */

[out:json];
(
  // get cycle route relatoins
  nwr
  [admin_level=7]
  ["ISO3166-2"~"MK"]
  [!"wikidata"]
  ({{bbox}});
);

out body;
>;
out skel qt;
```

but before starting to edit those almost 80 geometries I wonder if you know a way to automate the task, maybe using the `ISO` identifier as a join field between the OSM and Wikidata info. 

What do you think?

jsanz commented:
I run this oneliner to get from the Overpass Turbo GeoJSON export a CSV, and then join it with the result to a wikidata [query](https://query.wikidata.org/#SELECT%20%3Fid%20%3Fiso_3166_2%20%3Flabel_en%0AWHERE%20%7B%0A%20%20%23%20List%20of%20regions%2C%20whose%20sub-regions%20we%20want.%0A%20%20VALUES%20%3Fentity%20%7B%20wd%3AQ221%20%7D%0A%0A%20%20%23%20P150%20%3D%20%22contains%20administrative%20territorial%20entity%22%0A%20%20%3Fentity%20wdt%3AP150%2a%20%3Fid%20.%0A%20%20%3Fid%20wdt%3AP300%20%3Fiso_3166_2%0A%0A%20%20OPTIONAL%20%7B%20%3Fid%20rdfs%3Alabel%20%3Flabel_en%20.%20FILTER%28LANG%28%3Flabel_en%29%20%3D%20%22en%22%29%20%7D%0A%7D).

```
$ cat export.geojson | sed -e "s/ISO3166-2/iso/" \
  | jq ".features[] | { osm_id: .id, iso_3166_2: .properties.iso, osm_name: .properties.\"name:en\" }" \
  | json2csv \
  | sed -e "/node.*/d" \
  > osm-macedonia.csv

$  csvjoin -c iso_3166_2 osm-macedonia.csv query.csv | sed -e "s/http\:\/\/www\.wikidata\.org\/entity\///" | csvlook
|-------------------+------------+--------------------------------------+----------+------------+-----------------------------------|
|  osm_id           | iso_3166_2 | osm_name                             | id       | iso_3166_2 | label_en                          |
|-------------------+------------+--------------------------------------+----------+------------+-----------------------------------|
|  relation/2454874 | MK-42      | Municipality of Kochani              | Q1819289 | MK-42      | Kočani Municipality               |
|  relation/2460481 | MK-70      | Municipality of Sopishte             | Q2282240 | MK-70      | Sopište Municipality              |
|  relation/2460539 | MK-74      | Municipality of Studenichani         | Q2088034 | MK-74      | Studeničani Municipality          |
|  relation/2460678 | MK-32      | Municipality of Zelenikovo           | Q9050283 | MK-32      | Zelenikovo Municipality           |
|  relation/2460743 | MK-59      | Municipality of Petrovec             | Q1018191 | MK-59      | Petrovec Municipality             |
|  relation/2460762 | MK-34      | Municipality of Ilinden              | Q657754  | MK-34      | Ilinden Municipality              |
|  relation/2460854 | MK-82      | Municipality of Chucher Sandevo      | Q629720  | MK-82      | Čučer-Sandevo Municipality        |
.....
```

So now I have a CSV with both the Wikidata and the OSM identifiers.
```

jsanz commented:
Using [`osmapi`](http://osmapi.metaodi.ch/) I've been able to add the property just to one relation on this changeset: https://www.openstreetmap.org/changeset/74676469

I will do the same for the rest of the relations in another changeset later today

nickpeihl commented:
@jsanz [I added OSM relation tags](https://www.wikidata.org/wiki/Special:Contributions/Nickpeihl) to many of the municipalities in Wikidata yesterday. I haven't checked if all are complete though.

I started to create a task to add wiki tags to OSM using the [Sophox Editor](https://wiki.openstreetmap.org/wiki/Sophox_Editor) by @nyurik. But I have not yet finished the query. 

Maybe we should write documentation for the methodologies that we develop. I am sure they will be useful in the future.

jsanz commented:
+1 to document this. 

Sophox expects OSM data to have the Wikidata identifier, so I guess you need a different query than the one we've been using. At this moment [the query](https://sophox.org/#%20%20%20%20%20%20%20%20%20%20%20%20%23defaultView%3AMapRegions%0A%20%20%20%20%20%20%20%20%20%20%20%20SELECT%0A%20%20%20%20%20%20%20%20%20%20%20%20%28if%28bound%28%3Fid2%29%2C%3Fid2%2C%3Fid%29%20as%20%3Fid%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Fiso_3166_2%20%3Flabel_en%20%3Flabel_mk%20%3Flabel_sq%20%3FwkidataOSM%0A%20%20%20%20%20%20%20%20%20%20%20%20WHERE%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20SERVICE%20%3Chttps%3A%2F%2Fquery.wikidata.org%2Fsparql%3E%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20Using%20nested%20query%20to%20ensure%20there%20is%20only%20one%20%3Fid2%20value%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7BSELECT%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Fid%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%28SAMPLE%28%3Fid2%29%20as%20%3Fid2%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%28SAMPLE%28%3Fiso_3166_2%29%20as%20%3Fiso_3166_2%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%28SAMPLE%28%3Flabel_en%29%20as%20%3Flabel_en%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%28SAMPLE%28%3Flabel_mk%29%20as%20%3Flabel_mk%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%28SAMPLE%28%3Flabel_sq%29%20as%20%3Flabel_sq%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20WHERE%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20List%20of%20regions%2C%20whose%20sub-regions%20we%20want.%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20VALUES%20%3Fentity%20%7B%20wd%3AQ221%20%7D%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20P150%20%3D%20%22contains%20administrative%20territorial%20entity%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Fentity%20wdt%3AP150%2a%20%3Fid%20.%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3Fid%20wdt%3AP300%20%3Fiso_3166_2%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20OPTIONAL%20%7B%20%3Fid%20rdfs%3Alabel%20%3Flabel_en%20.%20FILTER%28LANG%28%3Flabel_en%29%20%3D%20%22en%22%29%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20OPTIONAL%20%7B%20%3Fid%20rdfs%3Alabel%20%3Flabel_mk%20.%20FILTER%28LANG%28%3Flabel_mk%29%20%3D%20%22mk%22%29%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20OPTIONAL%20%7B%20%3Fid%20rdfs%3Alabel%20%3Flabel_sq%20.%20FILTER%28LANG%28%3Flabel_sq%29%20%3D%20%22sq%22%29%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20remove%20possible%20ID%20duplicates%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20GROUP%20BY%20%3Fid%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20ORDER%20BY%20%3Fid%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D) only returns three geometries, the one that have the `wikidata` tag.

@nyurik told me yesterday to better add wikidata identifiers to OSM relations, so I will wait for confirmation from him to run this small python code to bulk modify the relationship:

```python
import osmapi 
import os 
import csv 
from pathlib import Path 
 
# Get creds from environment
osm_user = os.getenv('OSM_USER') 
osm_pass = os.getenv('OSM_PASS') 
 
assert osm_user and osm_pass 
 
api = osmapi.OsmApi(username=os.getenv('OSM_USER'),password=os.getenv('OSM_PASS')) 
 
# Get joined CSV (see PR comment)
csv_file = Path('./macedonia.csv') 
assert csv_file.exists() 
 
with csv_file.open('r') as csv_reader: 
    csv_dict = csv.DictReader(csv_reader) 
    api.ChangesetCreate({"comment": "adding wikidata property to Macedonia municipalities"}) 
     
    for row in csv_dict: 
        osm_id = row['osm_id'].split('/')[1]
        wd_id = row['id']
        assert osm_id and wd_id

        rel = api.RelationGet(osm_id)  
        rel['tag']['wikidata'] =  wd_id
         
        api.RelationUpdate(rel) 
    
    api.ChangesetClose()
```

nyurik commented:
Highly recommend not to use osmapi for these things. Usually JOSM offers most of the "bulk editing" capabilities, but they should be according to community guidelines.  Or lets sync and I can show you how to do it in Sophox as well.

jsanz commented:
Update: after @nyurik and @nickpeihl worked on generating a query to contribute to OSM the wikidata entries for North Macedonia municipalities, the standard query returns most of the geometries. Still this is a work in progress because we need to:

* Ensure the wikipedia pages linked in OSM relations are not disambiguation pages using a query similar to:

```
SELECT ?osmId ?wdLabel ?wd ?wp ?loc WHERE {
  # Limit to subjects that have a tag called "wikidata", and show its location
  ?osmId osmt:wikidata ?wd ;
         osmm:loc ?loc .

  # ?wd must be an "instance of" a disambiguation page, or an instance
  # of some type, which itself is a (sub-)*subclass of a disambig page.
  ?wd wdt:P31/wdt:P279* wd:Q4167410 .

  OPTIONAL { ?osmId osmt:wikipedia ?wp . }
            
  # Pick the first available language for the wikidata entry (creates ?wdLabel value)
  SERVICE wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE],en,de,fr,it,pl,ru,es,sv,nl" . }
}
```

* Filter those regions that are not official anymore, using this:
```
FILTER NOT EXISTS { ?id p:P300/pq:P582 ?end_time }
```
* MK-85, the capital region, is still missing. The [OSM relation](https://www.openstreetmap.org/relation/2439740) is identified, but in Wikidata is apparently [Q384](https://www.wikidata.org/wiki/Q384).

![image](https://user-images.githubusercontent.com/188264/65344749-d1282a00-dbd8-11e9-8b87-3a4936aca987.png)

nickpeihl commented:
The MK-85 identifier also assigned to [Q2575820](https://www.wikidata.org/wiki/Q2575820) (Greater Skopje). It seems to me that Skopje and Greater Skopje are different entities and Greater Skopje is most likely the matching entity to the ISO code. So perhaps Greater Skopje should have [OSM relation 2439740](https://www.openstreetmap.org/relation/2439740#map=11/42.0261/21.3700) and we remove that relation from [Q384](https://www.wikidata.org/wiki/Q384) (Skopje).

 @jsanz, what do you think?

jsanz commented:
@nickpeihl I've added the wikidata tag for the Great Skopje relation in OSM so now it points to  [Q2575820](https://www.wikidata.org/wiki/Q2575820). I also removed the OSM relation link for the city and placed it into the Greater Skopje entry.

Still, for the current query to return this area we need to add it to the country entry, but apparently I don't have the right to do so (too few edits I guess). Can you try to add `Q2575820` as a region from the country (`Q221`)?

As a workaround, I went from the property type instead of the boundaries listed in the Country (query in detail section). This works after filtering deprecated ISO codes **and** former municipalities. Do you think is this a valid approach instead of adding Great Skopje to the country?

<details><summary>sparql sophox query</summary>

```sparql
#defaultView:MapRegions
SELECT 
  ?id
  ?iso_3166_2 
  ?label_en ?label_mk ?label_sq 
  WHERE {
  SERVICE <https://query.wikidata.org/sparql> {
    {
      SELECT 
        ?id 
        (SAMPLE(?iso_3166_2) AS ?iso_3166_2) 
        (REPLACE(SAMPLE(?label_en), " Municipality", "") AS ?label_en) 
        (REPLACE(SAMPLE(?label_mk), "Општина ", "")  AS ?label_mk) 
        (REPLACE(SAMPLE(?label_sq), "Komuna e ", "")  AS ?label_sq) 
      WHERE {
        ?id (wdt:P31/wdt:P279*) wd:Q3395432.
        ?id wdt:P300 ?iso_3166_2.
        OPTIONAL { ?id rdfs:label ?label_en. FILTER((LANG(?label_en)) = "en") }
        OPTIONAL { ?id rdfs:label ?label_mk. FILTER((LANG(?label_mk)) = "mk") }
        OPTIONAL { ?id rdfs:label ?label_sq. FILTER((LANG(?label_sq)) = "sq") }
        
        # Remove those with deprecated ISO code
        FILTER(NOT EXISTS { ?id (p:P300/pq:P582) ?end_time_iso. })
      }
      GROUP BY ?id
    }
    # Remove former municipalities
    FILTER(NOT EXISTS { ?id (p:P31/pq:P582) ?end_time_instance. })
  }
}
ORDER BY ?iso_3166_2
```
</details>

nickpeihl commented on sources/mk/macedonia-municipalities.hjson:
```diff
@@ -0,0 +1,136 @@
+
+                # P150 = "contains administrative territorial entity"
+                ?entity wdt:P150* ?id .
+                ?id wdt:P300 ?iso_3166_2
```
```suggestion
                ?id wdt:P300 ?iso_3166_2
                
                FILTER NOT EXISTS { ?id p:P300/pq:P582 ?end_time }
```

nickpeihl commented on sources/mk/macedonia-municipalities.hjson:
```diff
@@ -0,0 +1,136 @@
+            #defaultView:MapRegions
+            SELECT
+            (if(bound(?id2),?id2,?id) as ?id)
+            ?iso_3166_2 ?label_en ?label_mk ?label_sq ?wkidataOSM
```
I don't think we need this variable.

```suggestion
            ?iso_3166_2 ?label_en ?label_mk ?label_sq 
```

jsanz replied:
Yeah I spotted that one on the last query I shared on the comments of the PR, thanks!!

nickpeihl commented on sources/mk/macedonia-municipalities.hjson:
```diff
@@ -0,0 +1,136 @@
+      hr: Općine u Makedoniji
+      hy: Մակեդոնիայի Հանրապետության վարչական բաժանում
+      id: Munisipalitas di Makedonia Utara
+      ie: Liste de municipalitates de Macedonia
```
Sometimes these labels come from tables in Wikipedia (e.g. List of...). Not a big deal since Kibana is not (and probably never will be) localized to [Interlingue](https://en.wikipedia.org/wiki/Interlingue). I have already fixed this in Wikidata.

```suggestion
      ie: Municipalitates de Macedonia
```

nickpeihl commented on sources/mk/macedonia-municipalities.hjson:
```diff
@@ -0,0 +1,136 @@
+    ttl: 108000
+    weight: 0
+    createdAt: /* TBD */
+    id: /* TBD */
```
We should be able to add these `/* TBD */` entries now.

nickpeihl commented on sources/mk/macedonia-municipalities.hjson:
```diff
@@ -0,0 +1,136 @@
+      cs: Seznam opštin v Severní Makedonii
+      de: Opština in Nordmazedonien
+      el: Δήμοι της Βόρειας Μακεδονίας
+      en: North Macedonia Municipality
```
```suggestion
      en: Macedonia Municipalities
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+ems-file-service+pull-request/8/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+ems-file-service+pull-request/9/)
